### PR TITLE
Promote StatefulSetAutoDeletePVC to stable in 1.32

### PR DESCRIPTION
**Pull request by mattcary**
/kind feature

Promotes [kubernetes/enhancements#1847](https://github.com/kubernetes/enhancements/issues/1847) to stable.

```release-note
Promoted feature gate `StatefulSetAutoDeletePVC` from beta to stable.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: KEP: https://github.com/kubernetes/enhancements/issues/1847
- [Usage]: https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#persistentvolumeclaim-retention
```

### CODE REVIEW

**Comment by sftim** [comment]
Changelog suggestion
```diff
-Feature gate StatefulSetAutoDeletePVC promoted to stable
+Promoted feature gate `StatefulSetAutoDeletePVC` from beta to stable.
```

**Comment by mattcary** [comment]
> Changelog suggestion
> 
> ```diff
> -Feature gate StatefulSetAutoDeletePVC promoted to stable
> +Promoted feature gate `StatefulSetAutoDeletePVC` from beta to stable.
> ```

done

**Comment by mattcary** [comment]
/test pull-kubernetes-e2e-gce	

Seems to be an unrelated flake.

File: pkg/apis/apps/v1/defaults.go
Comment (soltysh) [inline pkg/apis/apps/v1/defaults.go:127]: You need to leave this as is, based on the new guidance, see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md#compatibility-versions

File: pkg/controller/statefulset/stateful_pod_control.go
Comment (soltysh) [inline pkg/controller/statefulset/stateful_pod_control.go:127]: Same as before, leave it as is until we drop the feature.

File: pkg/features/versioned_kube_features.go
Comment (soltysh) [inline pkg/features/versioned_kube_features.go:684]: Based on the new guideline (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md#compatibility-versions) we'll remove it only in 1.35.

**Review by soltysh — COMMENTED** [review]
@mattcary have a look at the guidelines (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md#compatibility-versions) and update the PR accordingly. If in doubt feel free to ping me on slack.

**Comment by mattcary** [comment]
> @mattcary have a look at the guidelines (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md#compatibility-versions) and update the PR accordingly. If in doubt feel free to ping me on slack.

Ah, I thought since the feature gate is being locked I should drop all references to it immediately. Thanks for the link, I see I was incorrect. Sorry for the confusion. I think that will make this PR a lot simpler... let me start working on that.

File: pkg/features/versioned_kube_features.go
Comment (mattcary) [reply pkg/features/versioned_kube_features.go]: done

**Comment by mattcary** [comment]
Pulled out all of the test changes, I think this is correct now?

**Comment by mattcary** [comment]
Oops, I have to leave the feature gate check in, but the tests can no longer flip the gate off as it's locked. I'll fix that.

**Comment by soltysh** [comment]
> Ah, I thought since the feature gate is being locked I should drop all references to it immediately. Thanks for the link, I see I was incorrect. Sorry for the confusion. I think that will make this PR a lot simpler... let me start working on that.

No worries, it's a recent change. 

> Pulled out all of the test changes, I think this is correct now?

Looking...

**Review by soltysh — APPROVED** [review]
I left minor comments around api and validation, but I'll defer to api-reviewers.

the controller changes lgtm
/lgtm
/approve

File: pkg/controller/statefulset/stateful_pod_control_test.go
Comment (mattcary) [reply pkg/controller/statefulset/stateful_pod_control_test.go:612]: Done, both tests in this file & the one in stateful_set_control_test.go.

**Comment by jpbetz** [comment]
I did a quick scan of the changes and this all looks reasonable.

cc @aaron-prindle @Jefftree for compatibility version sanity check.

This will require approval from api-reviewers. I see the api-review label is already set.

File: test/e2e/auth/projected_clustertrustbundle.go
Comment (Jefftree) [inline test/e2e/auth/projected_clustertrustbundle.go:59]: Curious on why this change? Were tests not okay before?

Comment (mattcary) [reply test/e2e/auth/projected_clustertrustbundle.go]: Hunh, I think this was a merge/rebase that I missed. thanks! should be fixed now.

**Comment by mattcary** [comment]
Sorry all, looks like a I did something wrong with the rebase, there is a merge commit in here somehow. I'm in process of sorting it out.

File: staging/src/k8s.io/api/apps/v1beta2/types.go
Comment (aaron-prindle) [inline staging/src/k8s.io/api/apps/v1beta2/types.go:198]: @Jefftree would renaming the field here be breaking for v1.32 users using --emulated-version=v1.31 with old configs using the beta feature gate?  Does storage versions account for this in any way?

Comment (mattcary) [reply staging/src/k8s.io/api/apps/v1beta2/types.go:198]: ah, this was an old typo that I fixed. The type isn't used anywhere in k/k, but if you want me to revert it in case it's being used elsewhere I can.

**Comment by mattcary** [comment]
> Sorry all, looks like a I did something wrong with the rebase, there is a merge commit in here somehow. I'm in process of sorting it out.

Okay, I think I sorted it out. Should be good to go.

File: staging/src/k8s.io/api/apps/v1beta2/types.go
Comment (Jefftree) [reply staging/src/k8s.io/api/apps/v1beta2/types.go:198]: yeah this field doesn't look to be used with the feature so it should be more of an API review question than compatibility version one.

Comment (aaron-prindle) [reply staging/src/k8s.io/api/apps/v1beta2/types.go:198]: I see thanks.  This change LGTM with respect to Compatibility Versions

**Comment by aaron-prindle** [comment]
LGTM for the changes here with respect to Compatibility Versions.  No user facing beta/GA APIs are modified/removed and bumping the feature gate is inline w/ Compatibility Versions guarantees.

**Comment by mattcary** [comment]
/test pull-kubernetes-unit

That test is passing locally, hopefully it was a flake

File: pkg/apis/apps/v1/defaults_test.go
Comment (soltysh) [reply pkg/apis/apps/v1/defaults_test.go:600]: I'd probably leave that variable `enablePVCDeletionPolicy` as it was before, then you'd have here this:
```golang
if !test.enablePVCDeletionPolicy {
    featuregatetesting.SetFeatureGateEmulationVersionDuringTest(t, utilfeature.DefaultFeatureGate, version.MustParse("1.31"))
}
featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.StatefulSetAutoDeletePVC, test.enablePVCDeletionPolicy)
```
That's what we've been using all over the places, and this will also mean less changes :wink:

File: pkg/apis/apps/v1beta1/defaults_test.go
Comment (soltysh) [inline pkg/apis/apps/v1beta1/defaults_test.go:563]: Same comment here.

File: pkg/apis/apps/v1beta2/defaults_test.go
Comment (soltysh) [inline pkg/apis/apps/v1beta2/defaults_test.go:534]: Ditto.

File: staging/src/k8s.io/api/apps/v1beta2/types.go
Comment (liggitt) [reply staging/src/k8s.io/api/apps/v1beta2/types.go:198]: go-level changes have no bearing on the content of the REST API

**Comment by liggitt** [comment]
/approve
for API doc changes

File: pkg/apis/apps/validation/validation_test.go
Comment (soltysh) [reply pkg/apis/apps/validation/validation_test.go:545]: We don't want to remove any of the gating until the final removal of the feature gate. The goal is to support emulating older versions, so we want both the production code to support both paths and tests to exercise both are still working. 

There's more details in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md#ga-features

File: staging/src/k8s.io/api/apps/v1/types.go
Comment (soltysh) [reply staging/src/k8s.io/api/apps/v1/types.go:255]: In the constants section, to be precise.

File: pkg/apis/apps/validation/validation_test.go
Comment (atiratree) [reply pkg/apis/apps/validation/validation_test.go:545]: Ack, makes sense. This bit should not be marked as resolved then.

Comment (soltysh) [reply pkg/apis/apps/validation/validation_test.go:545]: This still holds.

**Review by soltysh — COMMENTED** [review]
To summarize:
- validation_test.go - revert back the feature gate testing
- defaults_test.go (in all version) - about just using the pattern I've pointed in https://github.com/kubernetes/kubernetes/pull/128247#discussion_r1830955119
- v1/types.go - constants description

These are the remaining bits.

File: pkg/apis/apps/v1beta1/defaults_test.go
Comment (mattcary) [reply pkg/apis/apps/v1beta1/defaults_test.go:563]: Okay, that means that either we'd have enablePVCDeletionPolicy added to most of the test cases, which seems error prone, or we continue testing with the feature gate off, which means when it is removed the tests will change as well.

I would have thought we would want most tests to be for what's feature-locked.

Comment (soltysh) [reply pkg/apis/apps/v1beta1/defaults_test.go:563]: I haven't thought about that, in that case, we can probably leave it in this direction, since that will be less changes.

Comment (mattcary) [reply pkg/apis/apps/v1beta1/defaults_test.go:563]: sgtm! thx

**Comment by knabben** [comment]
Hey @soltysh @mattcary 
⚠️ Do we still intend to merge this for v1.32? Just a reminder that the code freeze is starting 02:00 UTC Friday November 8th 2024 (a little less than 1 week from now). Please make sure the PR has both lgtm and approved labels before the code freeze. Thanks!

**Comment by mattcary** [comment]
> Hey @soltysh @mattcary ⚠️ Do we still intend to merge this for v1.32? Just a reminder that the code freeze is starting 02:00 UTC Friday November 8th 2024 (a little less than 1 week from now). Please make sure the PR has both lgtm and approved labels before the code freeze. Thanks!

Yup, @soltysh and I have been syncing on slack, there's a couple things we want that I should be able to do tonight and then this should be in tomorrow.

File: pkg/apis/apps/validation/validation_test.go
Comment (soltysh) [reply pkg/apis/apps/validation/validation_test.go:545]: I looked once more through the code, and I don't see us using the `StatefulSetAutoDeletePVC` feature gate at all in the validation code, which means it's not necessary.

**Comment by soltysh** [comment]
The current state is:
> validation_test.go - revert back the feature gate testing

Not needed, we're not using the `StatefulSetAutoDeletePVC` feature gate in validation at all.

> defaults_test.go (in all version) - about just using the pattern I've pointed in https://github.com/kubernetes/kubernetes/pull/128247#discussion_r1830955119

As discussed in that comment this is not needed.

> v1/types.go - constants description

This one still holds, so that's the only one left here to resolve.

File: pkg/apis/apps/validation/validation_test.go
Comment (soltysh) [reply pkg/apis/apps/validation/validation_test.go:545]: That's because we always want to validate these values, so dropping that test is totally fine here.

Comment (atiratree) [reply pkg/apis/apps/validation/validation_test.go:545]: Ok, cool.

**Comment by mattcary** [comment]
> The current state is:
> 
> > validation_test.go - revert back the feature gate testing
> 
> Not needed, we're not using the `StatefulSetAutoDeletePVC` feature gate in validation at all.
>
Ack
 
> > defaults_test.go (in all version) - about just using the pattern I've pointed in [#128247 (comment)](https://github.com/kubernetes/kubernetes/pull/128247#discussion_r1830955119)
> 
> As discussed in that comment this is not needed.
> 
Ack

> > v1/types.go - constants description
> 
> This one still holds, so that's the only one left here to resolve.

Done!

File: staging/src/k8s.io/api/apps/v1/types.go
Comment (mattcary) [reply staging/src/k8s.io/api/apps/v1/types.go:255]: done

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by mattcary** [comment]
/test pull-kubernetes-e2e-gce-cos-alpha-features 

seems to be a flake

**Pull request merged**
